File: src/communication/can_bridge.py
"""CAN-to-Ethernet bridge via TCP.

Connects to a CAN/Ethernet converter device and parses incoming
CAN frames into per-device payloads.  Frame format (binary)::

    device_id (2 bytes, big-endian)
    length    (1 byte)
    data      (N bytes)
    CRLF      (2 bytes, 0x0D 0x0A)
"""

from PyQt6.QtCore import QObject, pyqtSignal
from PyQt6.QtNetwork import QTcpSocket, QAbstractSocket
import struct
import logging

logger = logging.getLogger(__name__)


class CanBridge(QObject):
    """TCP client for a CAN/Ethernet converter box.

    The converter multiplexes multiple CAN devices over a single
    TCP connection.  Each inbound frame is tagged with a 2-byte
    *device_id* so that the bridge can demultiplex them.
    """

    can_frame_received = pyqtSignal(int, bytes)    # device_id, frame_data
    connection_state_changed = pyqtSignal(object)   # DeviceStatus or None

    def __init__(self, parent=None):
        super().__init__(parent)
        self._socket = QTcpSocket(self)
        self._buffer = bytearray()

        self._socket.connected.connect(self._on_connected)
        self._socket.disconnected.connect(self._on_disconnected)
        self._socket.readyRead.connect(self._on_ready_read)
        self._socket.errorOccurred.connect(self._on_error)

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def connect_to_converter(self, host: str, port: int):
        """Open a TCP connection to the CAN converter at *host*:*port*."""
        self._buffer.clear()
        self._socket.connectToHost(host, port)

    def disconnect_converter(self):
        """Close the TCP connection to the converter."""
        self._socket.disconnectFromHost()

    def is_connected(self) -> bool:
        """Return True if the socket is connected."""
        return (
            self._socket.state()
            == QAbstractSocket.SocketState.ConnectedState
        )

    # ------------------------------------------------------------------
    # Internal slots
    # ------------------------------------------------------------------

    def _on_connected(self):
        from ..core.common.types import DeviceStatus
        self.connection_state_changed.emit(DeviceStatus.IDLE)
        logger.info("CAN bridge connected")

    def _on_disconnected(self):
        from ..core.common.types import DeviceStatus
        self.connection_state_changed.emit(DeviceStatus.OFFLINE)
        logger.info("CAN bridge disconnected")

    def _on_error(self, error):
        logger.error(
            "CAN bridge socket error: %s", self._socket.errorString()
        )

    def _on_ready_read(self):
        data = self._socket.readAll().data()
        self._buffer.extend(data)
        # Frame format: device_id(2B BE) + length(1B) + data(NB) + CRLF
        while len(self._buffer) >= 5:
            device_id = struct.unpack('>H', self._buffer[0:2])[0]
            frame_len = self._buffer[2]
            total_len = 3 + frame_len + 2  # header + payload + CRLF
            if len(self._buffer) < total_len:
                break  # incomplete frame, wait for more data
            frame_data = bytes(self._buffer[3:3 + frame_len])
            self.can_frame_received.emit(device_id, frame_data)
            self._buffer = self._buffer[total_len:]

File: src/communication/can/CanBridge.cpp
#include "communication/can/CanBridge.h"
#include <QDebug>

CanBridge::CanBridge(QObject* parent)
    : QObject(parent)
    , m_socket(new QTcpSocket(this))
{
    connect(m_socket, &QTcpSocket::connected, this, &CanBridge::onConnected);
    connect(m_socket, &QTcpSocket::disconnected, this, &CanBridge::onDisconnected);
    connect(m_socket, &QTcpSocket::readyRead, this, &CanBridge::onReadyRead);
    connect(m_socket, &QTcpSocket::errorOccurred, this, &CanBridge::onSocketError);
}

CanBridge::~CanBridge()
{
    disconnectFromConverter();
}

void CanBridge::connectToConverter(const QString& host, int port)
{
    m_host = host;
    m_port = port;
    m_intentionalDisconnect = false;

    qDebug() << "CanBridge: Connecting to CAN converter at" << host << "port" << port;
    emit connectionStateChanged(DeviceStatus::Offline);
    m_socket->connectToHost(host, port);
}

void CanBridge::disconnectFromConverter()
{
    m_intentionalDisconnect = true;

    if (m_socket->state() != QAbstractSocket::UnconnectedState) {
        m_socket->disconnectFromHost();
    }
    m_readBuffer.clear();
}

bool CanBridge::isConnected() const
{
    return m_socket && m_socket->state() == QAbstractSocket::ConnectedState;
}

void CanBridge::onConnected()
{
    qDebug() << "CanBridge: Connected to CAN converter at" << m_host << ":" << m_port;
    m_readBuffer.clear();
    emit connectionStateChanged(DeviceStatus::Idle);
}

void CanBridge::onDisconnected()
{
    qDebug() << "CanBridge: Disconnected from CAN converter";
    if (!m_intentionalDisconnect) {
        emit connectionStateChanged(DeviceStatus::Offline);
    }
}

void CanBridge::onReadyRead()
{
    m_readBuffer.append(m_socket->readAll());
    parseFrames();
}

void CanBridge::onSocketError(QAbstractSocket::SocketError error)
{
    Q_UNUSED(error)
    qWarning() << "CanBridge: Socket error:" << m_socket->errorString();
}

void CanBridge::parseFrames()
{
    // CAN frame format from converter:
    //   Device ID (2 bytes, big-endian)
    //   Frame length (1 byte)
    //   Frame data (N bytes)
    //   Delimiter: 0x0D 0x0A (CRLF)

    static constexpr int HEADER_SIZE = 3; // 2 bytes device ID + 1 byte length
    static constexpr int DELIMITER_SIZE = 2; // CR + LF

    while (m_readBuffer.size() >= HEADER_SIZE) {
        // Look for the next complete frame
        // Minimum frame: HEADER_SIZE + 0 data + DELIMITER_SIZE = 5 bytes
        int frameLength = static_cast<unsigned char>(m_readBuffer[2]);
        int totalSize = HEADER_SIZE + frameLength + DELIMITER_SIZE;

        if (m_readBuffer.size() < totalSize) {
            // Not enough data yet, wait for more
            break;
        }

        // Verify delimiter
        if (static_cast<unsigned char>(m_readBuffer[HEADER_SIZE + frameLength]) != 0x0D ||
            static_cast<unsigned char>(m_readBuffer[HEADER_SIZE + frameLength + 1]) != 0x0A) {
            qWarning() << "CanBridge: Invalid frame delimiter, discarding 1 byte";
            m_readBuffer.remove(0, 1);
            continue;
        }

        // Extract device ID (big-endian, 2 bytes)
        int deviceId = (static_cast<unsigned char>(m_readBuffer[0]) << 8) |
                        static_cast<unsigned char>(m_readBuffer[1]);

        // Extract frame data
        QByteArray frameData = m_readBuffer.mid(HEADER_SIZE, frameLength);

        emit canFrameReceived(deviceId, frameData);

        // Remove the processed frame from the buffer
        m_readBuffer.remove(0, totalSize);
    }

    // Safety: if we have too much unparseable data, clear the buffer
    if (m_readBuffer.size() > 65536) {
        qWarning() << "CanBridge: Read buffer overflow, clearing" << m_readBuffer.size() << "bytes";
        m_readBuffer.clear();
    }
}
